test/core/xds/xds_client_test.cc

```diff
@@ -3779,6 +3803,367 @@ TEST_F(XdsClientTest, AdsReadWaitsForHandleRelease) {
+  TriggerConnectionFailure(fallback_server,
+                           absl::UnavailableError("Another server down"));
+  stream->MaybeSendStatusToClient(absl::UnavailableError("Fallback"));
+  auto error = watcher->WaitForNextError();
+  ASSERT_TRUE(error.has_value());
+  EXPECT_THAT(error->code(), absl::StatusCode::kUnavailable);
+  EXPECT_THAT(error->message(), ::testing::ContainsRegex("Another server down"))
+      << error->message();
+  error = watcher->WaitForNextError();
+  ASSERT_TRUE(error.has_value());
+  EXPECT_THAT(error->code(), absl::StatusCode::kUnavailable);
+  EXPECT_THAT(error->message(), ::testing::ContainsRegex("Fallback"))
+      << error->message();
+  EXPECT_THAT(GetServerConnections(),
+              ::testing::ElementsAre(
+                  ::testing::Pair(kDefaultXdsServerUrl, false),
+                  ::testing::Pair(fallback_server.server_uri(), false)));
+}
+
+TEST_F(XdsClientTest, FallbackOnStartup) {
+  FakeXdsBootstrap::FakeXdsServer primary_server;
+  FakeXdsBootstrap::FakeXdsServer fallback_server("fallback_xds_server");
+  // Regular operation
+  InitXdsClient(FakeXdsBootstrap::Builder().SetServers(
+      {primary_server, fallback_server}));
+  // Start a watch for "foo1".
+  auto watcher = StartFooWatch("foo1");
+  auto primary_stream = WaitForAdsStream(primary_server);
+  ASSERT_NE(primary_stream, nullptr);
+  // XdsClient should have sent a subscription request on the ADS stream.
+  auto request = WaitForRequest(primary_stream.get());
+  ASSERT_TRUE(request.has_value());
+  CheckRequest(*request, XdsFooResourceType::Get()->type_url(),
+               /*version_info=*/"", /*response_nonce=*/"",
+               /*error_detail=*/absl::OkStatus(),
+               /*resource_names=*/{"foo1"});
+  TriggerConnectionFailure(primary_server,
+                           absl::UnavailableError("Primary server is down"));
+  EXPECT_THAT(
+      GetServerConnections(),
+      ::testing::ElementsAre(::testing::Pair(kDefaultXdsServerUrl, false),
+                             ::testing::Pair("fallback_xds_server", true)));
+  // XdsClient should have created an ADS stream.
+  auto fallback_stream = WaitForAdsStream(fallback_server);
+  ASSERT_NE(fallback_stream, nullptr);
+  // XdsClient should have sent a subscription request on the ADS stream.
+  request = WaitForRequest(fallback_stream.get());
+  ASSERT_TRUE(request.has_value());
+  CheckRequest(*request, XdsFooResourceType::Get()->type_url(),
+               /*version_info=*/"", /*response_nonce=*/"",
+               /*error_detail=*/absl::OkStatus(),
+               /*resource_names=*/{"foo1"});
+  // Send a response.
+  fallback_stream->SendMessageToClient(
+      ResponseBuilder(XdsFooResourceType::Get()->type_url())
+          .set_version_info("1")
+          .set_nonce("A")
+          .AddFooResource(XdsFooResource("foo1", 6))
+          .Serialize());
+  // XdsClient should have delivered the response to the watcher.
```

Comment:
Please add a check here that `GetServerConnections()` shows a healthy connection to the fallback server.